trax/tf_numpy/examples/mnist/model.py

```diff
@@ -19,119 +19,109 @@
 
-    for w, b in zip(self.weights, self.biases):
-      x = sigmoid(np.dot(w, x.T).T + b)
-    return x
+        Args:
+          hidden_layers: List of ints specifying the sizes of hidden layers. Could
+            be empty.
+          input_size: Length of the input array. The network receives the input
+            image as a flattened 1-d array. Defaults to 784(28*28), the default
+            image size for MNIST.
+          num_classes: The number of output classes. Defaults to 10.
+        """
+        hidden_layers = [input_size] + hidden_layers + [num_classes]
+        self.params = []
+        for i in range(len(hidden_layers) - 1):
+            self.params.append(np.random.randn(hidden_layers[i + 1], hidden_layers[i]))
+            self.params.append(np.random.randn(hidden_layers[i + 1]))
 
-  def train(self, x, y, learning_rate=0.01):
-    """Runs a single training pass.
 
-    Args:
-      x: 2-d array of size batch_size x image_size.
-      y: 2-d array of size batch_size x num_classes in one-hot notation.
-      learning_rate: The learning rate.
-    """
-    x = np.array(x, copy=False)
-    y = np.array(y, copy=False)
+    def sigmoid(self, x):
+        return 1 / (1 + np.exp(-x))
 
-    def mean_squared_error(x, y):
-      diff = x - y
-      return np.sum(diff * diff) / len(x)
 
-    wb_tensors = [p.data for p in self.weights + self.biases]
-    with tf.GradientTape() as g:
-      g.watch(wb_tensors)
-      loss = mean_squared_error(self.forward(x), y)
-    gradients = g.gradient(loss.data, wb_tensors)
-    gradients = [np.asarray(grad) for grad in gradients]
+    def forward(self, x):
+        """Performs the forward pass.
 
-    new_weights_and_biases = []
-    for v, dv in zip(self.weights + self.biases, gradients):
-      new_weights_and_biases.append(v - learning_rate * dv)
+        Args:
+          x: 2-d array of size batch_size x image_size.
 
-    total_len = len(new_weights_and_biases)
-    self.weights = new_weights_and_biases[:total_len // 2]
-    self.biases = new_weights_and_biases[total_len // 2:]
+        Returns:
+          A 2-d array of size batch_size x num_classes.
+        """
 
-  def evaluate(self, x, y):
-    """Returns the number of correct predictions.
+        out = tf.cast(x, tf.float32)
 
-    Args:
-      x: 2-d array of size batch_size x image_size.
-      y: 2-d array of size batch_size x num_classes.
+        for i in range(0, len(self.params)-2, 2):
+            out = np.tanh(np.dot(out, self.params[i].T) + self.params[i+1])
 
-    Returns:
-      A scalar, the number of correct predictions.
-    """
-    y_actual = np.argmax(y, axis=1)
-    y_predicted = np.argmax(self.forward(x), axis=1)
-    return int(
-        np.sum(np.array(y_actual == y_predicted, copy=False, dtype=int32)))
+        out = self.sigmoid(np.dot(out, self.params[-2].T) + self.params[-1])
+
+        return out
+
+
+    def mean_squared_error(self, x, y):
+        y_out = self.forward(x)
```

Comment:
Move the self.forward into the train loop (so "self.mean_squared_error(self.forward(x), y)" below), so its clear that mean_squared_error is just calculating the error and not the forward pass as well.